### PR TITLE
JAMES-2530 Complete Rule DTO

### PR DESCRIPTION
**Pull request by chibenwa**
Based on https://github.com/linagora/james-project/pull/1679

### CODE REVIEW

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 1]
```diff
@@ -19,49 +19,318 @@
...
+        public static Action ofMailboxIds(List<String> mailboxIds) {
+            Preconditions.checkNotNull(mailboxIds, "mailboxIds should no be null");
+            return new Action(mailboxIds);
+        }
+
+        public static Action ofMailboxIds(String... mailboxIds) {
+            return ofMailboxIds(Arrays.asList(mailboxIds));
+        }
+
+        private final ImmutableList<String> mailboxIds;
```
Comment (trantienduchn) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: I though we don't need multi mailboxIds, from https://github.com/linagora/james-project/pull/1683#discussion_r212624616

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 2]
```diff
@@ -19,49 +19,318 @@
...
+                .add("comparator", comparator)
+                .add("value", value)
+                .toString();
+        }
     }
 
-    public Rule(String id) {
-        Preconditions.checkNotNull(id);
 
+    public static class Action {
```
Comment (trantienduchn) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: @rouazana sad in chat:
> maybe if you s/action/set the intent will be more understandable

https://chat.linagora.com/openpaas/pl/f8kzuez3738uuy1u93c7397ccw

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java (hunk 1 above)
Comment (rouazana) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: We don't, but we can keep it in the API and just forbit its use in our implementation.

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java (hunk 2 above)
Comment (rouazana) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: Yes, but matthieu's suggestion seems better, something like Action { Set { mailboxIds } }

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java (hunk 1 above)
Comment (chibenwa) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: I gonna throw in the JMAP layer for this.

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java (hunk 2 above)
Comment (chibenwa) [reply server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: I'll modify the POJO

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java [hunk 3]
```diff
@@ -47,13 +47,33 @@
     }
 
     public static RuleDTO from(Rule rule) {
-        return new RuleDTO(rule.getId());
+        return new RuleDTO(rule.getId().asString(),
+                rule.getName(),
+                rule.getCondition().getField().asString(),
+                rule.getCondition().getComparator().asString(),
+                rule.getCondition().getValue(),
+                rule.getAction().getAppendInMailboxes().getMailboxIds());
```
Comment (rouazana) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java]: you don't modify also the DTO?

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 4]
```diff
@@ -19,49 +19,355 @@
...
 import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
 
 public class Rule {
 
-    private final String id;
+    public static class Id {
+
+        public static Id of(String id) {
+            Preconditions.checkNotNull(id, "id should no be null");
```
Comment (rouazana) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: s/no/not

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 5]
```diff
@@ -19,49 +19,355 @@
...
+            Preconditions.checkArgument(StringUtils.isNotBlank(id), "id should no be empty");
```
Comment (rouazana) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: idem

File: server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java [hunk 6]
```diff
@@ -19,49 +19,355 @@
...
+            public String asString() {
+                return comparatorName;
+            }
+        }
+
+        public static Condition of(Field field, Comparator comparator, String value) {
+            Preconditions.checkNotNull(field, "field should no be null");
+            Preconditions.checkNotNull(comparator, "comparator should no be null");
+            Preconditions.checkNotNull(value, "value should no be null");
+            Preconditions.checkArgument(StringUtils.isNotBlank(value), "value should no be empty");
```
Comment (rouazana) [inline server/data/data-jmap/src/main/java/org/apache/james/jmap/api/filtering/Rule.java]: idem x4

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/RuleTest.java [hunk 7]
```diff
@@ -19,13 +19,25 @@
...
-class RuleTest {
+public class RuleTest {
+
+    private static final List<String> ACTION_MAILBOXIDS = Arrays.asList("id-01");
+    private static final String CONDITION_COMPARATOR = "contains";
+    private static final String CONDITION_FIELD = "cc";
+    private static final String NAME = "a name";
+    public static final Rule.Condition CONDITION = Rule.Condition.of(Rule.Condition.Field.of(CONDITION_FIELD), Rule.Condition.Comparator.of(CONDITION_COMPARATOR), "something");
+    public static final Rule.Action ACTION = Rule.Action.of(Rule.Action.AppendInMailboxes.withMailboxIds(ACTION_MAILBOXIDS));
+    public static final Rule.Id UNIQUE_ID = Rule.Id.of("uniqueId");
```
Comment (rouazana) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/RuleTest.java]: private for all 3?

File: server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/RuleTest.java [hunk 8]
```diff
@@ -34,8 +46,119 @@ void shouldMatchBeanContract() {
...
+    void innerClassActionShouldMatchBeanContract() {
+        EqualsVerifier.forClass(Rule.Action.class)
+            .verify();
+    }
+
+    @Test
+    void innerClassIdShouldMatchBeanContract() {
+        EqualsVerifier.forClass(Rule.Id.class)
+            .verify();
+    }
```
Comment (rouazana) [inline server/data/data-jmap/src/test/java/org/apache/james/jmap/api/filtering/RuleTest.java]: what about mailboxIds?

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java (hunk 3 above)
Comment (chibenwa) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java]: The DTI is *flat* so far. Should I give it a new shape to match the Rule object structure?

Comment (rouazana) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java]: Seems a good idea. Any reason to have it flat else?

Comment (chibenwa) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java]: 'cause I'm lazy?

Comment (chibenwa) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/filtering/RuleDTO.java]: Done

**Comment by trantienduchn** [comment]
my fixup is at https://github.com/linagora/james-project/pull/1683/commits/be839bc0988078c1b3de5a19971c7bb7f8384e70

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**